File: 백준/Silver/1874. 스택 수열/스택 수열.py
from collections import deque
import sys
ip, op = sys.stdin, sys.stdout
stk, ret = [], []
n = int(ip.readline())
temp = 0
flag = True
for _ in range(n):
    num = int(ip.readline())
    while temp < num:
        temp += 1
        stk.append(temp)
        ret.append('+')
    if stk and stk[-1] != num:
        flag = False
    elif stk and stk[-1] == num:
        stk.pop()
        ret.append('-')
print('\n'.join(ret) if flag else 'NO')

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        var in = new BufferedReader(new InputStreamReader(System.in));
        var out = new PrintWriter(System.out);

        int n = Integer.parseInt(in.readLine());
        int idx = 1;
        Deque<Integer> stk = new ArrayDeque<>();
        List<String> ret = new ArrayList<>();
        String plus = "+";
        String minus = "-";
        for(int i = 0; i < n; i++){
            int now = Integer.parseInt(in.readLine());
            while(idx <= now){
                stk.push(idx++);
                ret.add(plus);
            }
            if(!stk.isEmpty() && stk.peek() != now){
                System.out.println("NO");
                return;
            }
            if(!stk.isEmpty() && stk.peek() == now){
                stk.pop();
                ret.add(minus);
            }
        }
        for(String s : ret) out.println(s);



        out.flush();
        out.close();
    }
}